### ISSUE TITLE
Add List/Stream factory methods for primitive arrays

### ISSUE THREAD

**Issue by danieldietrich** [javaslang/javaslang#350]
* `of(boolean[])`
* `of(char[])`
* `of(double[])`
* `of(float[])`
* `of(int[])`
* `of(long[])`
* `of(short[])`

Resulting in

* `List<Boolean>` / `Stream<Boolean>`
* `List<Character>` / `Stream<Character>`
* ...

**Note:** Then we are able to convert a String into a List of digits

```java
// = List(1, 2, 3)
List.of("123".toCharArray()).map(c -> Character.digit(c, 10))
```

When collections evolve, also Set, ... should have these factory methods.

**Comment by danieldietrich** [comment]
There will be no ambiguities when naming the methods `of`, i.e. the following works:

```java
// = Stream.of(Integer...)
Stream<Integer> s1 = Stream.of(1, 2, 3);

// = Stream.of(int[])
Stream<Integer> s2 = Stream.of(new int[] {1, 2, 3});

// = Stream.of(Integer...)
Stream<Integer> s3 = Stream.of(new Integer[] {1, 2, 3});
```

**Comment by danieldietrich** [comment]
## Naming Things

We already have:

* `of(T)`
* `of(T...)`
* `ofAll(Iterable<T>)`
* `ofAll(Iterator<T>)`

Now, what is the best name for `of(int[])`, `of(long[])`, ...?

* `of` does not work because it introduces ambiguities (mentioned above).
* `ofAll` would emphasize that an array has multiple elements but `of(T...)` should then also be named `ofAll(T...)`. But this would differ from the Java standard API, e.g. `java.util.stream.Stream.of(T)`, `j.u.s.Stream.of(T...)`.

This leads to the decision to choose another name than `of` or `ofAll`.

Candidates are `ofInt` resp. `ofInts`.

* `ofInt` would emphasize that the _component type_ is `int` resp. `Integer`
* `ofInts` would emphasize that the _argument type_ is `int[]`

The all leads to a _brain loop_...

We could name things explicitely without introducing inconsistencies compared to Java's std lib, namely:

These will stay because of std lib conformity:

* `of(T)`
* `of(T...)`

These will change:

* `ofIterable(Iterable<T>)`
* `ofIterator(Iterator<T>)`

These will be new and also be consistent to the above:

* `ofInts(int[])`, `ofLongs(long[])`, ...

What I do not like about this is the _redundancy_ of the method names and the method arg type name (, i.e. `ofXxx(Xxx)`).

If this is a viable solution, the major 2.0.0 release would be the perfect moment to make slight API changes.

**Comment by danieldietrich** [comment]
Now we have:

```java
// = Stream.of(Integer...)
Stream<Integer> s1 = Stream.of(1, 2, 3);

// = Stream.of(int[])
Stream<Integer> s2 = Stream.ofInts(new int[]{1, 2, 3});

// = Stream.of(Integer...)
Stream<Integer> s3 = Stream.of(new Integer[]{1, 2, 3});

// Stream.of(T)
Stream<int[]> s4 = Stream.of(new int[]{1, 2, 3});

// Stream.of(T)
Stream<Integer[]> s5 = Stream.<Integer[]>of(new Integer[]{1, 2, 3});
```

**Comment by danieldietrich** [comment]
Also: remove `static <T> Stream<T> ofAll(Iterator<? extends T> elements)` completely. This is internal API and will be a private interface method in a future Java version.

Until this happens, calls to this method can be replaced by `ofAll(Iterable)`, i.e. `ofAll(() -> iterator)`

**Comment by danieldietrich** [comment]
I ask myself what happens if we just remove the `of(T)` method. Calling `of(T...)` with one argument should be equivalent. Or maybe additionally change `of(T...)` to `of(T, T...)` to enforce at least one argument. When an empty instance is needed we call `empty()` or `nil()`, depending on the type.

Given that, can we then call all factory methods `of()`? I will check this...

**Comment by danieldietrich** [comment]
An alternative (to `ofIterable()`, `ofInts()`, `ofLongs()`, ...) is to keep the `ofAll()` (also for primitive arrays). Then we have:

```java
// = Stream.of(T...)
Stream<Integer> s1 = Stream.of(1, 2, 3);

// = Stream.ofInts(int[])
Stream<Integer> s2 = Stream.ofAll(new int[] {1, 2, 3});

// = Stream.of(T...)
Stream<Integer> s3 = Stream.of(new Integer[] {1, 2, 3});

// Stream.of(T)
Stream<int[]> s4 = Stream.of(new int[] {1, 2, 3});

// Stream.of(T)
Stream<Integer[]> s5 = Stream.<Integer[]> of(new Integer[] {1, 2, 3});
```

**Note:** The above examples are not well ordered. This is better:

```java
Stream<Integer>   s1 = Stream.of(1, 2, 3);
                     = Stream.of(new Integer[] {1, 2, 3});
Stream<Integer[]> s2 = Stream.<Integer[]> of(new Integer[] {1, 2, 3});
Stream<int[]>     s3 = Stream.of(new int[] {1, 2, 3});
Stream<Integer>   s4 = Stream.ofAll(new int[] {1, 2, 3});
```

Here we see that this is crap from the user perspective.

Better would be to have just two methods:

* `of(T...)`
* `cons(T)`

The `cons(T)` factory method takes what comes - without regret:

```java
Stream<Integer[]> s1 = Stream.cons(new Integer[] {1, 2, 3});
Stream<int[]>     s2 = Stream.cons(new int[] {1, 2, 3});
```

The `of(T...)` factory method is _intelligent_ and checks the args to decide what to do:

```java
Stream<Integer> s3 = Stream.of(1, 2, 3);
Stream<Integer> s4 = Stream.of(new Integer[] {1, 2, 3});
Stream<Integer> s5 = Stream.of(new int[] {1, 2, 3});
Stream<Integer> s6 = Stream.of(List.of(1, 2, 3));
```

In other words the collection's `of(T...)` method decomposes varargs, Object arrays, primitive arrays and Iterables into its components.

If we need to construct a collection which contains exactly the given element, no matter which type, we need to use `cons(T)`.

Deal?

**Comment by danieldietrich** [comment]
This is the only viable solution:

```java
Stream<Integer>       s1 = Stream.of(1);
Stream<Integer>       s2 = Stream.of(1, 2, 3);
                      // = Stream.of(new Integer[] {1, 2, 3});
 
Stream<int[]>         s3 = Stream.of(new int[] {1, 2, 3});
Stream<List<Integer>> s4 = Stream.of(List.of(1, 2, 3));

Stream<Integer>       s5 = Stream.ofAll(new int[] {1, 2, 3});
Stream<Integer>       s6 = Stream.ofAll(List.of(1, 2, 3));
        
// cuckoo's egg
Stream<Integer[]>     s7 = Stream.<Integer[]> of(new Integer[]{1, 2, 3});
                      //!= Stream.<Integer[]> of(1, 2, 3);
```
